File: backend/app/websocket/manager.py
from typing import Any, Dict, Optional

from fastapi import WebSocket

from app.core.logging import logger
from app.utils.time import utc_now


class WebSocketManager:
    def __init__(self) -> None:
        self.connections: set[WebSocket] = set()

    async def connect(self, websocket: WebSocket) -> None:
        await websocket.accept()
        self.connections.add(websocket)

    def disconnect(self, websocket: WebSocket) -> None:
        self.connections.discard(websocket)

    async def broadcast(self, payload: Dict[str, Any]) -> None:
        dead: list[WebSocket] = []
        for conn in self.connections:
            try:
                await conn.send_json(payload)
            except RuntimeError:
                dead.append(conn)
        for conn in dead:
            self.connections.discard(conn)

    async def emit_log(
        self, level: str, message: str, meta: Optional[Dict[str, Any]] = None
    ) -> None:
        log_message = message.strip()
        if not log_message:
            return
        if level == "error":
            logger.error(log_message)
        elif level == "warn":
            logger.warning(log_message)
        else:
            logger.info(log_message)
        await self.broadcast(
            {
                "type": "log",
                "level": level,
                "message": log_message,
                "timestamp": utc_now(),
                "meta": meta,
            }
        )


manager = WebSocketManager()

File: backend-bun/src/websocket/manager.ts
import type { ServerWebSocket } from "bun";
import { utcNow } from "../lib/time.ts";

export interface WSData {
  id: string;
}

class WebSocketManager {
  private connections = new Set<ServerWebSocket<WSData>>();

  connect(ws: ServerWebSocket<WSData>): void {
    this.connections.add(ws);
  }

  disconnect(ws: ServerWebSocket<WSData>): void {
    this.connections.delete(ws);
  }

  broadcast(payload: Record<string, unknown>): void {
    const message = JSON.stringify(payload);
    const dead: ServerWebSocket<WSData>[] = [];

    for (const conn of this.connections) {
      try {
        conn.send(message);
      } catch {
        dead.push(conn);
      }
    }

    for (const conn of dead) {
      this.connections.delete(conn);
    }
  }

  emitLog(level: string, message: string, meta?: Record<string, unknown>): void {
    const logMessage = message.trim();
    if (!logMessage) return;

    // Log to console
    if (level === "error") {
      console.error(logMessage);
    } else if (level === "warn") {
      console.warn(logMessage);
    } else {
      console.log(logMessage);
    }

    // Broadcast to connected clients
    this.broadcast({
      type: "log",
      level,
      message: logMessage,
      timestamp: utcNow(),
      meta,
    });
  }

  emitJobStatus(jobId: string, status: string): void {
    this.broadcast({
      type: "job.status",
      job_id: jobId,
      status,
    });
  }

  emitJobProgress(jobId: string, progress: number): void {
    this.broadcast({
      type: "job.progress",
      job_id: jobId,
      progress,
    });
  }

  get connectionCount(): number {
    return this.connections.size;
  }
}

export const manager = new WebSocketManager();
